lib/puppet/provider/user/windows_adsi.rb

```diff
@@ -143,6 +393,54 @@ def uid=(value)
     fail "uid is read-only"
   end
 
+  def attributes
+    attributes_hash = {}
+    managed_attributes.each do |attribute, info|
+      attributes_hash[attribute] = info[:getter].call(self) 
+    end
+
+    attributes_hash
+  end
+
+  def validate_attributes(new_attributes)
+    # Check for any unmanaged attributes
+
+    unmanaged_attributes = new_attributes.keys.reject do |attribute|
+      managed_attributes.keys.include?(attribute)
+    end
+
+    unless unmanaged_attributes.empty?
+      raise ArgumentError, _("Cannot manage the %{unmanaged_attributes} attributes. The manageable attributes are %{managed_attributes}.") % { unmanaged_attributes: unmanaged_attributes.join(', '), managed_attributes: managed_attributes.keys.join(', ') }
+    end
+
+    # Check for any invalid attribute combinations
+
+    if new_attributes[:password_change_not_allowed] == 'true' &&
+       new_attributes[:password_change_required]    == 'true'
+      raise ArgumentError, _("Cannot have password_change_not_allowed == true and password_change_required == true. Reason: Disallowing the user to change their password while also requiring them to change it the next time they log-in is a contradiction!")
+    end
+
+    if new_attributes[:password_change_required] == 'true' &&
+       new_attributes[:password_never_expires]   == 'true'
+      raise ArgumentError, _("Cannot have password_change_required == true and password_never_expires == true! Reason: password_change_required is enforced by immediately expiring the password. For a password that never expires, Windows will throw an 'Access Denied' error when you try to change the password upon the first login.")
+    end
+  end
+
+  def attributes=(new_attributes)
```

Comment:
Note that the changes here will _not_ be committed until flush is called or when the password's changed, which is consistent with the other properties in the provider. Is this fine?